control-plane-operator/hostedclusterconfigoperator/controllers/hcpstatus/hcpstatus.go

```diff
@@ -74,76 +74,57 @@ func (h *hcpStatusReconciler) Reconcile(ctx context.Context, req reconcile.Reque
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
 
-	log := ctrl.LoggerFrom(ctx)
+	return nil
+}
 
+func (h *hcpStatusReconciler) reconcile(ctx context.Context, hcp *hyperv1.HostedControlPlane) error {
+	log := ctrl.LoggerFrom(ctx)
 	log.Info("Reconciling hosted cluster version conditions")
+
 	var clusterVersion configv1.ClusterVersion
 	err := h.hostedClusterClient.Get(ctx, crclient.ObjectKey{Name: "version"}, &clusterVersion)
-	failingCondition := func() metav1.Condition {
-		if err != nil {
-			return metav1.Condition{
-				Type:    string(hyperv1.ClusterVersionFailing),
-				Status:  metav1.ConditionUnknown,
-				Reason:  hyperv1.ClusterVersionStatusUnknownReason,
-				Message: fmt.Sprintf("failed to get clusterversion: %v", err),
-			}
-		}
-		message := ""
-		for _, cond := range clusterVersion.Status.Conditions {
-			if cond.Type == "Failing" {
-				if cond.Status == configv1.ConditionTrue {
-					return metav1.Condition{
-						Type:    string(hyperv1.ClusterVersionFailing),
-						Status:  metav1.ConditionTrue,
-						Reason:  cond.Reason,
-						Message: cond.Message,
-					}
-				}
-			}
-			if cond.Type == "Progressing" {
-				message = cond.Message
-			}
-		}
-		return metav1.Condition{
-			Type:    string(hyperv1.ClusterVersionFailing),
-			Status:  metav1.ConditionFalse,
-			Reason:  hyperv1.AsExpectedReason,
-			Message: message,
-		}
-	}()
-	upgradeableCondition := func() metav1.Condition {
-		if err != nil {
-			return metav1.Condition{
-				Type:    string(hyperv1.ClusterVersionUpgradeable),
-				Status:  metav1.ConditionUnknown,
-				Reason:  hyperv1.ClusterVersionStatusUnknownReason,
-				Message: fmt.Sprintf("failed to get clusterversion: %v", err),
-			}
+	// We check err in lopp below to build conditions with ConditionUnknown status for all types.
+
+	cvoConditions := map[hyperv1.ConditionType]*configv1.ClusterOperatorStatusCondition{
+		hyperv1.ClusterVersionFailing:         findClusterOperatorStatusCondition(clusterVersion.Status.Conditions, "Failing"),
+		hyperv1.ClusterVersionProgressing:     findClusterOperatorStatusCondition(clusterVersion.Status.Conditions, "Progressing"),
+		hyperv1.ClusterVersionReleaseAccepted: findClusterOperatorStatusCondition(clusterVersion.Status.Conditions, "ReleaseAccepted"),
+		hyperv1.ClusterVersionUpgradeable:     findClusterOperatorStatusCondition(clusterVersion.Status.Conditions, configv1.OperatorUpgradeable),
+	}
+
+	for conditionType := range cvoConditions {
+		var hcpCVOCondition metav1.Condition
+		// Set unknown status.
+		hcpCVOCondition = metav1.Condition{
+			Type:               string(conditionType),
+			Status:             metav1.ConditionUnknown,
+			Reason:             hyperv1.ClusterVersionStatusUnknownReason,
+			Message:            fmt.Sprintf("failed to get clusterversion: %v", err),
```

Comment:
Looks like this message is not correct if err == nil and the condition type is not yet set